File: 0004 - Median of Two Sorted Arrays/solution.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        m, n = len(nums1), len(nums2)
        m1 = self.getKth(nums1, 0, m-1, nums2, 0, n-1, (m+n-1)//2)
        m2 = self.getKth(nums1, 0, m-1, nums2, 0, n-1, (m+n)//2)
        return (m1+m2)/2
    def getKth(self, n1, s1, e1, n2, s2, e2, k):
        if s1>e1: return n2[s2+k]
        if s2>e2: return n1[s1+k]
        i1, i2 = s1+(e1-s1+1)//2, s2+(e2-s2+1)//2
        if (i1-s1)+(i2-s2)<k:
            if n1[i1]<n2[i2]: return self.getKth(n1, i1+1, e1, n2, s2, e2, k-(i1-s1)-1)
            else: return self.getKth(n1, s1, e1, n2, i2+1, e2, k-(i2-s2)-1)
        else:
            if n1[i1]<n2[i2]: return self.getKth(n1, s1, e1, n2, s2, i2-1, k)
            else: return self.getKth(n1, s1, i1-1, n2, s2, e2, k)

File: 0004 - Median of Two Sorted Arrays/solution.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int n1 = nums1.size(), n2 = nums2.size();
        if(n1 > n2) return findMedianSortedArrays(nums2, nums1);
        int l = 0, r = 2*n1;
        while(l <= r){
            int mid1 = l+(r-l)/2, mid2 = n1+n2-mid1;
            int l1 = mid1==0?INT_MIN:nums1[(mid1-1)/2];
            int l2 = mid2==0?INT_MIN:nums2[(mid2-1)/2];
            int r1 = mid1==2*n1?INT_MAX:nums1[mid1/2];
            int r2 = mid2==2*n2?INT_MAX:nums2[mid2/2];
            if(l2 > r1) l = mid1+1;
            else if(l1 > r2) r = mid1-1;
            else return (max(l1, l2)+min(r1, r2))/2.0;
        }
        return -1;
    }
};